packages/next/src/server/stream-utils/node-web-streams-helper.ts

```diff
@@ -193,6 +193,89 @@ export function renderToInitialFizzStream({
   )
 }
 
+function createMetadataTransformStream(
+  insert: () => Promise<string> | string
+): TransformStream<Uint8Array, Uint8Array> {
+  let chunkIndex = -1
+  let isMarkRemoved = false
+
+  return new TransformStream({
+    async transform(chunk, controller) {
+      let iconMarkIndex = -1
+      let closedHeadIndex = -1
+      chunkIndex++
+
+      if (isMarkRemoved) {
+        controller.enqueue(chunk)
+        return
+      }
+      let iconMarkLength = 0
+      // Only search for the closed head tag once
+      if (iconMarkIndex === -1) {
+        iconMarkIndex = indexOfUint8Array(chunk, ENCODED_TAGS.META.ICON_MARK)
+        if (iconMarkIndex === -1) {
+          controller.enqueue(chunk)
+          return
+        } else {
+          // When we found the `<meta name="«nxt-icon»"` tag prefix, we will remove it from the chunk.
+          // Its close tag could either be `/>` or `>`, checking the next char to ensure we cover both cases.
+          iconMarkLength = ENCODED_TAGS.META.ICON_MARK.length
+          // Check if next char is /, this is for xml mode.
+          if (chunk[iconMarkIndex + iconMarkLength] === 47) {
+            iconMarkLength += 2
+          } else {
+            // The last char is `>`
+            iconMarkLength++
+          }
+        }
+      }
+
+      // Check if icon mark is inside <head> tag in the first chunk.
+      if (chunkIndex === 0) {
+        closedHeadIndex = indexOfUint8Array(chunk, ENCODED_TAGS.CLOSED.HEAD)
+        // The mark icon is located in the 1st chunk before the head tag.
+        // We do not need to insert the script tag in this case because it's in the head.
+        // Just remove the icon mark from the chunk.
+        if (iconMarkIndex < closedHeadIndex && iconMarkIndex !== -1) {
+          const replaced = new Uint8Array(chunk.length - iconMarkLength)
+
+          // Remove the icon mark from the chunk.
+          replaced.set(chunk.subarray(0, iconMarkIndex))
+          replaced.set(
+            chunk.subarray(iconMarkIndex + iconMarkLength),
+            iconMarkIndex
+          )
+          chunk = replaced
+          isMarkRemoved = true
+        }
+      } else {
```

Comment:
I think we can make the assumption that if the mark is in the first chunk it is also inside the head. This is b/c the mark is part of the shell and React will have hoisted it to the head if it is in the shell. So you can skip checking for the head tag and just do the substitution